server/src/main/java/org/apache/druid/curator/discovery/CuratorDruidLeaderSelector.java

```diff
@@ -215,4 +208,37 @@ public void unregisterListener()
     CloseableUtils.closeAndSuppressExceptions(leaderLatch.get(), e -> log.warn(e, "Failed to close LeaderLatch."));
     listenerExecutor.shutdownNow();
   }
+
+  // Method to handle connection state changes
+  private void handleConnectionStateChanged(CuratorFramework client, ConnectionState newState)
+  {
+    switch (newState) {
+      case SUSPENDED:
+      case LOST:
+        recreateLeaderLatch();
+        break;
+      case RECONNECTED:
+        // Connection reestablished, no action needed here
+        break;
+      default:
+        // Do nothing for other states
+        break;
+    }
+  }
+
+  private void recreateLeaderLatch()
+  {
+    // Close existing leader latch
+    CloseableUtils.closeAndSuppressExceptions(leaderLatch.get(), e -> log.warn(e, "Failed to close LeaderLatch."));
+
+    // Create and start a new leader latch
+    LeaderLatch newLeaderLatch = createNewLeaderLatchWithListener();
```

Comment:
The latch returned by `createNewLeaderLatchWithListener()` is actually the old latch.

Reply:
Re-wrote the method per feedback [here](https://github.com/apache/druid/pull/16528#discussion_r1622747033)